contributors/design-proposals/scheduling/scheduling-framework.md

```diff
@@ -1,436 +1,620 @@
+
+When relevant, plugin evaluation order is specified by the order the plugins
+appear in the configuration. A plugin that registers for multiple extension
+points can have different ordering at each extension point.
+
+### Optional Args
+
+Plugins may receive arguments from their config with arbitrary structure.
+Because one plugin may appear in multiple extension points, the config is in a
+separate list of `PluginConfig`.
+
+For example,
+
+```json
+{
+   "name": "ServiceAffinity",
+   "args": {
+      "LabelName": "app",
+      "LabelValue": "mysql"
+   }
+}
+```
+
+```go
+func NewServiceAffinity(args json.RawMessage, h PluginHandle) (Plugin, error) {
+    var config struct {
+        LabelName, LabelValue string
+    }
+    if err := json.Unmarshal(args, &config); err != nil {
+        return nil, errors.Wrap(err, "could not parse args")
+    }
+    //...
+}
+```
+
+### Backward compatibility
+
+The current `KubeSchedulerConfiguration` kind has `apiVersion:
+kubescheduler.config.k8s.io/v1alpha1`. This new config format will be either
+`v1alpha2` or `v1beta1`. When a newer version of the scheduler parses a
+`v1alpha1`, the "policy" section will be used to construct an equivalent plugin
+configuration.
+
+*Note: Moving `KubeSchedulerConfiguration` to `v1` is outside the scope of this
+design, but see also
+https://github.com/kubernetes/enhancements/blob/master/keps/sig-cluster-lifecycle/0032-create-a-k8s-io-component-repo.md
+and https://github.com/kubernetes/community/pull/3008*
+
+# USE CASES
+
+These are just a few examples of how the scheduling framework can be used.
+
+## Coscheduling
+
+Functionality similar to
+[kube-batch](https://github.com/kubernetes-sigs/kube-batch) (sometimes called
+"gang scheduling") could be implemented as a plugin. For pods in a batch, the
+plugin would "accumulate" pods in the [permit](#permit) phase by using the
+"wait" option. Because the permit stage happens after [reserve](#reserve),
+subsequent pods will be scheduled as if the waiting pod is using those
+resources. Once enough pods from the batch are waiting, they can all be
+approved.
+
+## Dynamic Resource Binding
+
+[Topology-Aware Volume Provisioning](https://kubernetes.io/blog/2018/10/11/topology-aware-volume-provisioning-in-kubernetes/)
+can be (re)implemented as a plugin that registers for [filter](#filter) and
+[pre-bind](#pre-bind) extension points. At the filtering phase, the plugin can
+ensure that the pod will be scheduled in a zone which is capable of provisioning
+the desired volume. Then at the pre-bind phase, the plugin can provision the
+volume before letting scheduler bind the pod.
+
+## Custom Scheduler Plugins (out of tree)
+
+The scheduling framework allows people to write custom, performant scheduler
+features without forking the scheduler's code. To accomplish this, developers
+just need to write their own `main()` wrapper around the scheduler. Because
+plugins must be compiled with the scheduler, writing a wrapper around `main()`
```

Comment:
if the plugins must be compiled with the scheduler , then it would keep growing the kubernetes in tree code or am i missing something ?

Reply:
For out of tree plugins, they would not add to the amount of code that sig-scheduling has to maintain. This design also has the goal that custom scheduler authors can vendor the standard scheduler implementation without modifying any of that code. Should make it much easier to update deps and recompile.

plugins that belong in the standard scheduler will be in our tree, increasing the total line count. However, those lines of code can be added with zero changes to the scheduling "core".